### PR TITLE
simplify the generated parse tree

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**
This addresses #60 by removing unnecessary groups (`atom` and `parentheses`, whose eval functions just returned the group children) and by conditionally grouping sum/product/parallel/power/negate only if there are children.

This generates much cleaner parse trees. For example, `3^2+(2)` generates

```xml
<sum>
  <power>
    <number>
      <ITEM>3</ITEM>
    </number>
    <number>
      <ITEM>2</ITEM>
    </number>
  </power>
  <ITEM>+</ITEM>
  <number>
    <ITEM>2</ITEM>
  </number>
</sum>
```

whereas the old version would generate:

```xml          
<sum>
  <product>
    <parallel>
      <negation>
        <power>
          <atom>
            <number>
              <ITEM>3</ITEM>
            </number>
          </atom>
          <atom>
            <number>
              <ITEM>2</ITEM>
            </number>
          </atom>
        </power>
      </negation>
    </parallel>
  </product>
  <ITEM>+</ITEM>
  <product>
    <parallel>
      <negation>
        <power>
          <atom>
            <parentheses>
              <sum>
                <product>
                  <parallel>
                    <negation>
                      <power>
                        <atom>
                          <number>
                            <ITEM>2</ITEM>
                          </number>
                        </atom>
                      </power>
                    </negation>
                  </parallel>
                </product>
              </sum>
            </parentheses>
          </atom>
        </power>
      </negation>
    </parallel>
  </product>
</sum>
```

### CODE REVIEW

**Comment by ChristopherChudzicki** [comment]
@jolyonb What do you think about this?

There is an argument to be made that the original version better represents what our parser is actually doing, and so might be more useful for debugging. But the original version is also unwieldy.

Note: I also doubt this will improve performance significantly if at all. It's the parsing that is slow, and I really haven't changed the parsing much, just what is stored.

The only sense in which this might be faster is that before, each evaluation call had to deal with a bunch of sum/diff/power that had exactly 1 child. Now those only-child cases are dealt with at parse time instead of at evaluation time. But they were probably fast to deal with in the first place.

**Comment by jolyonb** [comment]
This is actually a really clever way to do it. I completely approve. Some minor nits incoming.

File: mitxgraders/helpers/calc.py [hunk 1]
```diff
@@ -275,6 +272,20 @@ def vector_parse_action(self, tokens):
         """
         self.vectors_used = True
 
+    @staticmethod
+    def group_if_multiple(name):
+        """
+        Generates a parse action that groups ParseResults with given name if
+        ParseREsults has multiple children.
```
Comment (jolyonb) [inline mitxgraders/helpers/calc.py]: `ParseResults`, not `ParseREsults`. I told you there were minor nits!

File: mitxgraders/helpers/calc.py [hunk 2]
```diff
@@ -363,9 +374,7 @@ def parse_algebra(self):
         function.setParseAction(self.function_parse_action)
 
         # Define parentheses
-        parentheses = Group(Suppress("(") +
-                            expr +
-                            Suppress(")"))("parentheses")
+        parentheses = Suppress("(") + expr + Suppress(")")
```
Comment (jolyonb) [inline mitxgraders/helpers/calc.py]: Can we actually keep parentheses around, please? I like having these with an explicit grouping tag in debug mode.

File: mitxgraders/helpers/calc.py [hunk 3]
```diff
@@ -610,22 +621,6 @@ def eval_function(self, parse_result):
                        "Its input does not seem to be in its domain.").format(name=name)
                 raise FunctionEvalError(msg)
 
-    def eval_parens(self, parse_result):
```
Comment (jolyonb) [inline mitxgraders/helpers/calc.py]: Honestly, this can be made into a lambda function in the dictionary above.

**Comment by jolyonb** [comment]
`eval_arguments` and `eval_var` should also probably just be lambda functions, given that they're one-liners.

**Comment by ChristopherChudzicki** [comment]
@jolyonb I restored the parentheses. I don't mind having simple things as named functions, then they show up in tracebacks nicer. But for some reason I did make eval_parentheses a lambda. I am inconsistent. 

Anyway, you can merge or lambdify those other ones if you want :)

**Comment by jolyonb** [comment]
Haha, ok ;-) Very nice. Merging now!

**Pull request merged**